// simple-jwt-auth/dist/cjs/getPublicKey/extractPublicKeyFromJwksUri.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const error_fns_1 = require("@ehmpathy/error-fns");
const axios_1 = __importDefault(require("axios"));
const test_fns_1 = require("test-fns");
const extractPublicKeyFromJwksUri_1 = require("./extractPublicKeyFromJwksUri");
jest.mock('axios');
const axiosGetMock = axios_1.default.get;
const realExampleJwks = [
    {
        kty: 'RSA',
        n: 'rwDJOPJ7rzu2oaixavxpfMiSr4Th_bKnMQqykmQYC6IdY-A2RMseGSaQqxpX7SBKtoFqp0WjLbD1aj2iHH2i1nFQLNXxStwTbaJiwYiB_Bpsm1KLuk0oYnIn9AwEwiGlRNTWxJFqyNI3Z-XTZakmQCx71_QQ6hL6vfx-ensNJjgCUJB9Yz-lBkF3esYLcyc2SbndshYZ_qBj-AUb-8JzILfHBQ4kXGPhiV2063RCWWBynML9y-VC1bQF0lzAuyP4CX8IdHqlzck2zD11eHIIK2Bnz8C0DhRldabsfXaGREw4SQNqadyfpXgPUL1blcQNp7IZE_l3B8Zf0yTxRgh4mCWoe1hECgzOsfuaeeBY15vBO7otLSQGYejo8q5V90fk6RpOyqeRsM0KswqqTpgNUsyvuTWzDGfTTJtKnuItcn51C4r-_oyV7ICDYbjpid8Ejmy9ABe0ndrVb5rpDI6MzbiJhfeMF5nNQMyGrEqXRzBgwIybmO3f9FMeTwwOAG0LawM3pDwJf5lldj88x9edyWF7RemWGi68JF3PGVGbDXyWtu-cL7mvecvH37x1dmHsUWRxmWm6VTEozhfdENYDmOSl0FV5xmWXEsd-SttOC1cuGZQI66EaPMck9Z_7vqJT84TCm7zsOYBQStXM6wuTA3TnCSBa4IiUvQwuNcuhiSU',
        e: 'AQAB',
        kid: '4d.c71b8fd1-bba7-47ee-966b-65ab85b34972',
    },
];
const realExampleAlternateJwks = { keys: realExampleJwks };
describe('extractPublicKeyFromJwksUri', () => {
    beforeEach(() => jest.clearAllMocks());
    (0, test_fns_1.given)('a resolvable jwks uri', () => {
        (0, test_fns_1.given)('the body is the array of keys directly', () => {
            beforeEach(() => axiosGetMock.mockReturnValue({ data: realExampleJwks }));
            (0, test_fns_1.when)('the jwks has the kid of the token', () => {
                (0, test_fns_1.then)('it should return the public key of that key', () => __awaiter(void 0, void 0, void 0, function* () {
                    var _a;
                    const publicKey = yield (0, extractPublicKeyFromJwksUri_1.extractPublicKeyFromJwksUri)({
                        jwksUri: '__jwks_uri__',
                        headerClaims: {
                            kid: (_a = realExampleJwks[0]) === null || _a === void 0 ? void 0 : _a.kid,
                        },
                    });
                    expect(publicKey).toContain('-----BEGIN PUBLIC KEY-----');
                }));
            });
            (0, test_fns_1.when)('the jwks does not include the kid of the token', () => {
                (0, test_fns_1.then)('it should throw an error reporting this', () => __awaiter(void 0, void 0, void 0, function* () {
                    const error = yield (0, error_fns_1.getError)((0, extractPublicKeyFromJwksUri_1.extractPublicKeyFromJwksUri)({
                        jwksUri: '__jwks_uri__',
                        headerClaims: {
                            kid: '__not_to_be_found__',
                        },
                    }));
                    expect(error).toBeInstanceOf(extractPublicKeyFromJwksUri_1.ExtractPublicKeyFromJwksUriError);
                    expect(error.message).toContain('Could not find a JSON Web Key (JWK) with the KeyId specified by the token ');
                }));
            });
        });
        (0, test_fns_1.given)('the body is an object with the jwks nested under the .keys property', () => {
            beforeEach(() => axiosGetMock.mockReturnValue({ data: realExampleAlternateJwks }));
            (0, test_fns_1.when)('the jwks has the kid of the token', () => {
                (0, test_fns_1.then)('it should return the public key of that key', () => __awaiter(void 0, void 0, void 0, function* () {
                    var _a;
                    const publicKey = yield (0, extractPublicKeyFromJwksUri_1.extractPublicKeyFromJwksUri)({
                        jwksUri: '__jwks_uri__',
                        headerClaims: {
                            kid: (_a = realExampleJwks[0]) === null || _a === void 0 ? void 0 : _a.kid,
                        },
                    });
                    expect(publicKey).toContain('-----BEGIN PUBLIC KEY-----');
                }));
            });
        });
    });
});
