// simple-jwt-auth/dist/cjs/createSecureDistributedAuthToken.integration.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const test_fns_1 = require("test-fns");
const createSecureDistributedAuthToken_1 = require("./createSecureDistributedAuthToken");
const getAuthedClaims_1 = require("./getAuthedClaims");
const getPublicKey_1 = require("./getPublicKey/getPublicKey");
const createVerifiableSignature_test_1 = require("./signingAlgorithm/createVerifiableSignature.test");
// mock that we discover the real public key each time
jest.mock('./getPublicKey/getPublicKey');
const getPublicKeyMock = getPublicKey_1.getPublicKey;
describe('createSecureToken', () => {
    (0, test_fns_1.given)('rsa signing algorithm', () => {
        const alg = 'RS256';
        const keypair = createVerifiableSignature_test_1.exampleRsaKeyPair;
        beforeEach(() => {
            getPublicKeyMock.mockReturnValue(keypair.publicKey);
        });
        it('should be able to create a token that we can later getAuthedClaims on', () => __awaiter(void 0, void 0, void 0, function* () {
            // create a token
            const token = (0, createSecureDistributedAuthToken_1.createSecureDistributedAuthToken)({
                headerClaims: { alg, kid: '4.some_directory', typ: 'JWT' },
                claims: {
                    iss: 'https://auth.whodis.io/...',
                    aud: '__some_directory__',
                    sub: '__some_user__',
                    exp: 2516239022,
                },
                privateKey: keypair.privateKey,
            });
            expect(typeof token).toEqual('string'); // sanity check
            // check that we can auth on it, if the publicKey is discoverable
            const claims = yield (0, getAuthedClaims_1.getAuthedClaims)({
                token,
                issuer: 'https://auth.whodis.io/...',
                audience: '__some_directory__',
            });
            expect(claims).toEqual({
                iss: 'https://auth.whodis.io/...',
                aud: '__some_directory__',
                sub: '__some_user__',
                exp: 2516239022,
            });
        }));
    });
    (0, test_fns_1.given)('ecc signing algorithm', () => {
        const alg = 'ES256';
        const keypair = createVerifiableSignature_test_1.exampleEccKeyPair;
        beforeEach(() => {
            getPublicKeyMock.mockReturnValue(keypair.publicKey);
        });
        it('should be able to create a token that we can later getAuthedClaims on', () => __awaiter(void 0, void 0, void 0, function* () {
            // create a token
            const token = (0, createSecureDistributedAuthToken_1.createSecureDistributedAuthToken)({
                headerClaims: { alg, kid: '4.some_directory', typ: 'JWT' },
                claims: {
                    iss: 'https://auth.whodis.io/...',
                    aud: '__some_directory__',
                    sub: '__some_user__',
                    exp: 2516239022,
                },
                privateKey: keypair.privateKey,
            });
            expect(typeof token).toEqual('string'); // sanity check
            // check that we can auth on it, if the publicKey is discoverable
            const claims = yield (0, getAuthedClaims_1.getAuthedClaims)({
                token,
                issuer: 'https://auth.whodis.io/...',
                audience: '__some_directory__',
            });
            expect(claims).toEqual({
                iss: 'https://auth.whodis.io/...',
                aud: '__some_directory__',
                sub: '__some_user__',
                exp: 2516239022,
            });
        }));
    });
});
